### DISCUSSION TITLE
Memory consumption extremely high?

### DISCUSSION THREAD

**Discussion by Weissrolf** [libjxl/libjxl#3497] [Q&A]
First of all: I did not test libjxl directly, but through its implementation in Irfanview and XnViewMP.

**Describe the bug**
Unfortunately memory consumption goes through the roof for JXL encoding and decoding.

Encoding a 19200 x 25200 px image of 1.35 gb uncompressed TIFF size uses more than my 16 gb for encoding and thus needs to make heavy use of the pagefile to finish.

Decoding the image peaks at over 12 gb, while decoding the same as TIFF or PNG only hits a peak of about 1.35 gb and JPEG a peak of about 2.3 gb. 

**To Reproduce**
Encode/Decode (uncompressed) 1.35 gb large image file (19200 x 25200 px).

**Expected behavior**
Should not use over 10 times the memory size of the original image to encode, maybe 2-3 times max. Should not use close to 10 times the memory size to decode.

**Screenshots**
https://i.imgur.com/oB6Nc6N.png

**Environment**
 - OS: Windows 11
 - CPU type: 5900X

**Comment by Fraetor** [comment]
Encoding is quite memory intensive with default settings. Lower efforts/higher speeds have a much reduced memory usage:
See https://gitlab.com/wg1/jpeg-xl/-/issues/99 for more details.

**Comment by Weissrolf** [comment]
Isn't speed 7 the default? Even at speed 1 it hits the ceiling, though, just needs a bit less pagefile swapping on a 16 gb system.

And *de*coding these images peaks at over 13 gb (!) memory usage, too. Meanwhile JPEG manages with less than 2.5 gb and TIFF/PNG mostly just peak corresponding to the image size (about 1.4 gb).

I might have to check with CLI tools instead of Irfanview/XnViewMP for a better comparison.

**Comment by jonsneyers** [comment]
Both cjxl encode and djxl decode use full float buffers to represent images. That already means 12 or 16 bytes per pixel instead of 3 or 4 bytes per pixel, i.e. 4x the uncompressed size (if the original is 8-bit). On top of that, the encoder is storing all the ANS tokens (for histogram computation/context clustering and because encode is LIFO in ANS), which in case of lossless is another 8 bytes per token, i.e. 24 or 32 bytes per pixel. So for 8-bit images, lossless libjxl encode will need at least 12x the uncompressed image size (the good news is that for 16-bit images, it's 'only' 6x and for 32-bit images, it's 'only' 3x). A specialized encoder that is optimized for low bit depth (say at most 14-bit) could probably easily cut that in half. An encoder that doesn't try to do global optimization could do things in constant memory, at the cost of some compression density.

If lossless compression is what you're after, you can try the `fjxl` encoder, which should use no more than the uncompressed buffer size plus some small constant overhead. It only works for 8-bit RGBA images, but that's likely OK for your use case.

As for the decoder: The API does support decoding to non-full buffers (pixel callback that gives you a row of a tile at a time), or decoding to uint8 or uint16 buffers instead of float ones, so it's certainly possible to avoid such high memory consumption — this is what the chrome and firefox integrations do at the moment.

**Comment by Weissrolf** [comment]
Thanks for the explanation! My numbers were for _lossy_ compression on 8 bit RGB (non A) images, though (squeezing that 1.35 gb image down to 25-35 mb depending on speed settings).

I cannot fully test the Chrome integration, because Chrome refuses to display the test image at all, it turns to light gray from medium gray once it's finished decoding, but it doesn't show anything except the solid gray. During "decoding" Chrome peaks at around 3.9 gb for the 1.35 gb image, which is less compared to decoding a PNG version, but _double_ as much compared to decoding a JPG version. I will have to install the FF nightly to test that one.

Is there a technical reason why cjxl and djxl use full float buffers even for 8 bit images? Are websites supposed to fallback to JPEG or WebP for mobile devices with limited memory then?

**Comment by jonsneyers** [comment]
Doing everything with full float buffers is convenient for implementation, but it does quadruple the memory needed compared to using uint8. We already have decode options that are more memory-friendly (you can decode to uint8 buffers or use the callback to decode even without a full frame buffer).

Chrome not showing the test image could be a bug in chrome: could you share the image that is a valid jxl but chrome doesn't show it properly?

cjxl and djxl are just example applications so they're not really aiming for being memory optimized. It would be useful to have encode options in the api to reduce memory needs, but this hasn't been a priority to implement so far.

**Reply by MasterInQuestion** [reply]
For most cases JXL is actually handling 8bpc (bit-per-component).
    So the 32-bit float actually becomes wasteful and ridiculous excess...
    And leading to real issues:

    [ https://github.com/libjxl/libjxl/issues/464#issuecomment-903690594](https://github.com/libjxl/libjxl/issues/464#issuecomment-903690594)
    [ https://github.com/libjxl/libjxl/issues/1378#issuecomment-1110249317](https://github.com/libjxl/libjxl/issues/1378#issuecomment-1110249317)
    [ https://github.com/libjxl/libjxl/issues/1475#issuecomment-1146755204](https://github.com/libjxl/libjxl/issues/1475#issuecomment-1146755204)
    [ https://github.com/libjxl/libjxl/issues/1737#issuecomment-1662311297](https://github.com/libjxl/libjxl/issues/1737#issuecomment-1662311297)
    (so frequently several tens GiB level of encoding memory requirement..?)

**Comment by Weissrolf** [comment]
Here is the test image:

https://filetransfer.io/data-package/CgpuieAc#link

**Comment by cgohlke** [comment]
FWIW, the excessive peak memory usage during encoding can be reproduced using libjxl 0.6.1 via the Python [imagecodecs](https://pypi.org/project/imagecodecs/) package. Decoding to uint8 buffer looks reasonable:

```Python
import psutil
from imagecodecs import jpegxl_decode, jpegxl_encode

with open('briarstone_asylumt1X.jxl', 'rb') as fh:
    image = jpegxl_decode(fh.read())
print(f'image size: {image.size / 2**30:.3f} GB')
print(f'decode: {psutil.Process().memory_info().peak_wset / 2**30:.3f} GB')
encoded = jpegxl_encode(image)
print(f'encode: {psutil.Process().memory_info().peak_wset / 2**30:.3f} GB')
```
Output on Python 3.9 for Windows 10:
```
image size: 1.352 GB
decode: 2.430 GB
encode: 29.943 GB
```

**Comment by jonsneyers** [comment]
Things should have improved significantly since version 0.10.

Moving this to Discussions now.

**Comment by j7nj7n** [comment]
Lossless JPEG transcoding with -j 1 still uses an extreme amount of memory (on win64 with cjxl.exe). 94 megapixels, 50 MB uses over 5 gigs of RAM. Encoding with -j 0 uses only about 800 MB. The program should do piece-wise processing of the input when not aiming for maximum compression ratio to be scalable for large images. Perhaps this could be linked to the effort parameter.
